Gen typed params and latest resources: autogen java spec a5eee30

mickjermsurawong-stripe opened the pull request:
This is the latest updates that introduced typed params and new resources.
Jar from this is also used to run codes to verify api sample refs are up-to-date
cc @stripe/api-libraries

remi-stripe commented on src/main/java/com/stripe/param/AccountCreateParams.java:
```diff
@@ -0,0 +1,2642 @@
+   * dictionary containing a user's bank account details.
+   */
+  @SerializedName("bank_account")
+  Object bankAccount;
```
Why is this one an Object when `Company` is a class. I'm assuming it's because it has different shapes but what does the code using it looks like?

Also, I'm realizing that `bank_account` should not be supported at all. Only `external_account` should be supported. `bank_account` was removed in 2015 in theory: https://stripe.com/docs/upgrades#2015-10-01

mickjermsurawong-stripe replied:
This is removed and now we have `externalAccount` String token instead.
The documentation is still wrong though. I have a PR for gated docs that doesn't change that in API refs (saying that it's still a hash)

remi-stripe commented on src/main/java/com/stripe/param/AccountCreateParams.java:
```diff
@@ -0,0 +1,2642 @@
+
+  /** Specifies which fields in the response should be expanded. */
+  @SerializedName("expand")
+  List<String> expand;
```
Can you confirm whether we're always going to have this property on all params even if nothing is expandable?

mickjermsurawong-stripe replied:
ahh thank you remi! no. we should not expect this. I will have to fix this in OpenAPI spec.

remi-stripe replied:
It is likely easier to just have it always. Though I'm fine with not having it.

mickjermsurawong-stripe replied:
Yup we currently have it as default except request with `DELETE` endpoint

remi-stripe commented on src/main/java/com/stripe/param/AccountCreateParams.java:
```diff
@@ -0,0 +1,2642 @@
+   * have some of its requested capabilities be active and some be inactive.
+   */
+  @SerializedName("requested_capabilities")
+  List<RequestedCapability> requestedCapabilities;
```
I did not know we were going with enums in params. Is that expected? While I'm fine in principle, I would say that enum on resources are higher priority than on params (personal opinion). This is a bad example as it's not returned in the API but I've confirmed that `disabled_reason` for example on the `Account` is not an enum.

mickjermsurawong-stripe replied:
Yup this is expected. Enum on resource introduced now will cause too much incompatibility with 7.x. But here since we start from zero, we introduce Enum as well.

remi-stripe replied:
Ack okay. Does Java let you pass a string there easily instead of an enum value? There are a lot of properties that have gated enum values

mickjermsurawong-stripe replied:
We have `.toMap` exposed to users so they can always add gated/beta content the Map, and use the old way of calling API

remi-stripe commented on src/main/java/com/stripe/param/AccountCreateParams.java:
```diff
@@ -0,0 +1,2642 @@
+
+    /** Whether information has been collected from the company's directors. */
+    @SerializedName("directors_provided")
+    Boolean directorsProvided;
```
Random but wanted to make sure we won't have the same issue as we had in go for years where this sends `false` if it's not initialized. We had to move to pointers and that was a lot of work.

mickjermsurawong-stripe replied:
Added test to verify this f8ede6b

mickjermsurawong-stripe commented on src/main/java/com/stripe/param/AccountCreateParams.java:
```diff
@@ -0,0 +1,2642 @@
+     * Either a token, like the ones returned by [Stripe.js](https://stripe.com/docs/stripe.js), or
+     * a dictionary containing a user's bank account details.
+     */
+    public Builder setBankAccount(Map<String, Object> bankAccount) {
```
@remi-stripe when a field is polymorphic, the builder will have explicit setters for each type. So using the builder, one will have this effect of `setBankAccount` showing explicit options for all the possible types.
![image](https://user-images.githubusercontent.com/42550941/53468942-025cc500-3a11-11e9-8573-74d2b5ba9ffa.png)

The getter of the built object though will have return type Object that users have to cast, but that's not the expected use case.

remi-stripe replied:
Ack okay. I got confused because I expected the String or Class of params for bank accounts like @ob-stripe did in .net with AnyOf

remi-stripe commented on src/main/java/com/stripe/param/CustomerCreateParams.java:
```diff
@@ -0,0 +1,611 @@
+      return this;
+    }
+
+    public Builder setSource(Map<String, Object> source) {
```
I don't think we should support a map here. In theory, we should go all the way and support passing the correct list of params that are valid for the `source` hash like card's number, cvc, etc.
If we do, there's a big documentation risk as we need to make it clear no one should use this.

mickjermsurawong-stripe replied:
Ah we don't have this encoded in OpenAPI spec for now.. Let me see if I can fix it upstream..

remi-stripe replied:
Yeah this is one of those where I'm ambivalent. I want us to support it properly. But then more and more merchants will send raw PANs because it looks easy in the library. Not sure what the right product decision is here. Maybe the hash is easier because less people would find it.

mickjermsurawong-stripe replied:
As discussed, it's a now a String

remi-stripe commented on src/main/java/com/stripe/param/SubscriptionCreateParams.java:
```diff
@@ -0,0 +1,758 @@
+   * month for subsequent invoices.
+   */
+  @SerializedName("billing_cycle_anchor")
+  String billingCycleAnchor;
```
This should be `long` right? since it's a unix timestamp

mickjermsurawong-stripe replied:
@remi-stripe, I realized I haven't fixed this type inconsistencies

remi-stripe commented on src/main/java/com/stripe/param/SubscriptionCreateParams.java:
```diff
@@ -0,0 +1,758 @@
+   * trial immediately.
+   */
+  @SerializedName("trial_end")
+  String trialEnd;
```
This should be a string for `now` or a `long` for unix timestamp. Would it be cleaner to handle this the right way?

mickjermsurawong-stripe replied:
below here too

mickjermsurawong-stripe commented on src/main/java/com/stripe/param/CustomerCreateParams.java:
```diff
@@ -0,0 +1,611 @@
+
+    private Map<String, String> metadata;
+
+    private Object shipping;
```
Do you mean the `EMPTY` enum for `setShipping`?

remi-stripe replied:
Yes. This feels quite weird and I noticed it in other places. I guess we might not have a choice. Just looks quite hacky/wrong to me so wanted to raise.

mickjermsurawong-stripe replied:
We actually discussed this during the implementation phase, and agreed on `EMPTY` 
https://paper.dropbox.com/doc/Polymorphic-params-in-stripe-java--AZOgnte1Z7stWBn8JIdFsIe4Ag-amkaRe4C2QWqjVWcONtWx

remi-stripe commented on src/main/java/com/stripe/param/SubscriptionCreateParams.java:
```diff
@@ -0,0 +1,758 @@
+   */
+  @SerializedName("metadata")
+  Map<String, String> metadata;
+
```
We're missing `plan` though I guess it's on purpose because we consider it deprecated and we're fine forcing merchants to not support it? (it sounds okay especially as there's always the escape hatch anyway, just asking to confirm)

mickjermsurawong-stripe replied:
@remi-stripe I still haven't addressed this "missing" `plan` field and also Customer create too.
Are we okay with releasing without first?

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2527 @@
+  Account account;
+
+  @SerializedName("amount")
+  Long amount;
```
This should not be in this API. Please make sure to remove/document in pay-server so that it can't leak again. This is a really old and sensitive parameter that lets you do card testing really easily :(

mickjermsurawong-stripe replied:
This and the token fields below are addressed

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2527 @@
+  String customer;
+
+  @SerializedName("email")
+  String email;
```
AFAIK this is useful only for Legacy Checkout in go. This should not be in the library

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2527 @@
+  List<String> expand;
+
+  @SerializedName("external_id")
+  String externalId;
```
This is internal/sensitive field used by our SDKs like Stripe.js and should not leak in the library

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2527 @@
+  String externalId;
+
+  @SerializedName("iovation_blackbox")
+  String iovationBlackbox;
```
This is internal/sensitive field used by our SDKs like Stripe.js and should not leak in the library

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2527 @@
+  String iovationBlackbox;
+
+  @SerializedName("ip")
+  String ip;
```
This is internal/sensitive field used by our SDKs like Stripe.js and should not leak in the library

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2527 @@
+  String ip;
+
+  @SerializedName("payment_user_agent")
+  String paymentUserAgent;
```
This is internal/sensitive field used by our SDKs like Stripe.js and should not leak in the library

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2527 @@
+  Pii pii;
+
+  @SerializedName("recipient")
+  String recipient;
```
We likely should remove recipient from this, it's been deprecated for years

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2527 @@
+  String recipient;
+
+  @SerializedName("referrer")
+  String referrer;
```
This is internal/sensitive field used by our SDKs like Stripe.js and should not leak in the library

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2527 @@
+  String referrer;
+
+  @SerializedName("request_id")
+  String requestId;
```
I don't understand where that one is coming from but it should not be a parameter at all.

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2527 @@
+  String requestId;
+
+  @SerializedName("user_agent")
+  String userAgent;
```
This is internal/sensitive field used by our SDKs like Stripe.js and should not leak in the library

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2527 @@
+  String userAgent;
+
+  @SerializedName("validation_type")
+  ValidationType validationType;
```
Same as `amount`, can be used for card-testing, we need to explicitly hide this.

mickjermsurawong-stripe commented:
Thanks Remi!

On tokens there those internal fields, I will have to undoc them from public spec.

On the deprecated fields that are introduced, like `bank_account` on Account create, and `customer/id` on Customer create, they are the "forever_changes" version change; we remove from the public doc, but continue supporting. In OpenAPI if I understand correctly, those are now included for param log redaction, and also for internal testing. So they should be removed from SDK spec only.

So I think we likely want to ignore this forever change in SDK spec, and follow the latest params instead similar to that in API refs instead. That's a little tricky because the API refs also does lots of custom handling for these fields..

mickjermsurawong-stripe commented on src/main/java/com/stripe/param/CardUpdateOnCustomerParams.java:
```diff
@@ -0,0 +1,448 @@
+
+@Getter
+public class CardUpdateOnCustomerParams extends ApiRequestParams {
+  /** The name of the person or business that owns the bank account. */
```
Updated. Card params now have only card-related values

mickjermsurawong-stripe commented on src/main/java/com/stripe/param/CardUpdateOnCustomerParams.java:
```diff
@@ -0,0 +1,448 @@
+import lombok.Getter;
+
+@Getter
+public class CardUpdateOnCustomerParams extends ApiRequestParams {
```
I suspect the pattern of usage is from the method call where they look at method signature, instead of going through our list of params. 
Do you think naming like `CustomerCardUpdate` and `AccountCardUpdate` would be better? We can do `CustomerCardUpdate` and `PayoutCardUpdate` but that will be more confusing.

mickjermsurawong-stripe commented on src/main/java/com/stripe/param/CardUpdateOnCustomerParams.java:
```diff
@@ -0,0 +1,448 @@
+
+  /** The type of entity that holds the account. This can be either `individual` or `company`. */
+  @SerializedName("account_holder_type")
+  AccountHolderType accountHolderType;
```
Addressed

mickjermsurawong-stripe commented on src/main/java/com/stripe/param/CustomerCreateParams.java:
```diff
@@ -0,0 +1,611 @@
+  String coupon;
+
+  @SerializedName("customer")
+  String customer;
```
Addressed

mickjermsurawong-stripe commented on src/main/java/com/stripe/param/CustomerCreateParams.java:
```diff
@@ -0,0 +1,611 @@
+
+  /** A custom ID to use for the customer. */
+  @SerializedName("id")
+  String id;
```
It now goes away together with `customer`

mickjermsurawong-stripe commented:
@remi-stripe 
I've addressed all the comments except one on typing of `billingCycleAnchor`, `trialEnd` in `SubscriptionCreateParams.java`. I'll look at how to resolve that.

Could you help to look at a few more critical params please?

mickjermsurawong-stripe commented:
ptal @remi-stripe

remi-stripe commented on src/main/java/com/stripe/param/AccountCreateParams.java:
```diff
@@ -0,0 +1,2587 @@
+  String externalAccount;
+
+  @SerializedName("from_recipient")
+  String fromRecipient;
```
This has been undocumented for a while, why is it appearing here when others undocumented things don't appear?

mickjermsurawong-stripe replied:
addressed in https://github.com/stripe/stripe-java/pull/711/commits/760b5b5713915de21ce7368dace7e88e9c5dc85f#diff-a9f0113cf10a4a96bc61507995cc5a57R62

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2512 @@
+      return this;
+    }
+
+    public Builder setCard(EncryptedCard card) {
```
This is secret and gated and I don't think it should be here

mickjermsurawong-stripe replied:
addressed

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2512 @@
+      return this;
+    }
+
+    public Builder setCard(ApplePay card) {
```
I don't think this should be in stripe-java. If it's what I think it's about creating a card token from an Apple Pay PKToken and should only be in stripe-ios

mickjermsurawong-stripe replied:
addressed

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2512 @@
+      return this;
+    }
+
+    public Builder setCard(CreditCard card) {
```
CreditCard is a risky name for a class like this. It implies Debit cards would not work

mickjermsurawong-stripe replied:
Addressed

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2512 @@
+      return this;
+    }
+
+    public Builder setCard(SwipeData card) {
```
I don't think we should support that publicly.

mickjermsurawong-stripe replied:
Addressed

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2512 @@
+    String customer;
+
+    @SerializedName("default_for_currency")
+    Boolean defaultForCurrency;
```
I don't think this is a real thing on token creation so I don't get why it's here

mickjermsurawong-stripe replied:
Addressed

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2512 @@
+    String number;
+
+    @SerializedName("number_looks_valid")
+    Boolean numberLooksValid;
```
same, this looks like really leaking something

mickjermsurawong-stripe replied:
Addressed

remi-stripe commented on src/main/java/com/stripe/param/TokenCreateParams.java:
```diff
@@ -0,0 +1,2512 @@
+       */
+      public Builder putMetadata(String key, String value) {
+        if (this.metadata == null) {
+          t
```
This should not be here, there's no 3DS class on token creation and if there is it's some weird internal thing I would say.

mickjermsurawong-stripe replied:
Removed 3d secured

remi-stripe commented on src/main/java/com/stripe/param/SourceCreateParams.java:
```diff
@@ -0,0 +1,1105 @@
+
+    public enum NotificationMethod implements ApiRequestParams.Enum {
+      @SerializedName("deprecated_none")
+      DEPRECATED_NONE("deprecated_none"),
```
Can you clarify what this one even is?

mickjermsurawong-stripe replied:
This goes away when we decide to undoc mandate

remi-stripe commented on src/main/java/com/stripe/param/SourceCreateParams.java:
```diff
@@ -0,0 +1,1105 @@
+      DEPRECATED_NONE("deprecated_none"),
+
+      @SerializedName("email")
+      EMAIL("email"),
```
What is that `EMAIL()` thing

mickjermsurawong-stripe replied:
Same because of undoc mandate
